### ISSUE TITLE
Unexpected allocation in logical operators

### ISSUE THREAD

**Issue by dolejm1** [JuliaImages/Images.jl#386]
There is a lot off allocations in the function:
function ensureConectivity1{T,N}(labels::Image{T,N,Array{T,N}})
        nlabels = copy(labels)
        nlabel = one(T)
        for i_2 = 1: size(labels, 2)
            for i_1 = 1: size(labels, 1)
                if (zero(T) < nlabels[i_1, i_2])
                    nlabels[i_1, i_2] = nlabel
                end
            end
        end
        return nlabels, nlabel - one(T)
    end

but if I change the nlabels in the is less statement to labels, all the allocations vanish
The changed function is:
function ensureConectivity1{T,N}(labels::Image{T,N,Array{T,N}})
        nlabels = copy(labels)
        nlabel = one(T)
        for i_2 = 1: size(labels, 2)
            for i_1 = 1: size(labels, 1)
                if (zero(T) < labels[i_1, i_2])
                    nlabels[i_1, i_2] = nlabel
                end
            end
        end
        return nlabels, nlabel - one(T)
    end

I don't see the allocations if the function operates on Array instead of Image:
function ensureConectivity1{T,N}(labels::Array{T,N})
        nlabels = copy(labels)
        nlabel = one(T)
        for i_2 = 1: size(labels, 2)
            for i_1 = 1: size(labels, 1)
                if (zero(T) < nlabels[i_1, i_2])
                    nlabels[i_1, i_2] = nlabel
                end
            end
        end
        return nlabels, nlabel - one(T)
    end

**Comment by timholy** [comment]
This comes down to a julia bug, unfortunately. This would be fixed by #370, but I had to revert that because of https://github.com/JuliaLang/julia/issues/13327.

Probably your best fix is to have the version for images call a version for arrays:

```jl
ensureConnectivity1(img::AbstractImageDirect) = copyproperties(img, ensureConnectivity1(data(img)))

function ensureConnectivity(labels::AbstractMatrix)
    # code goes here
end
```
(Given your code, you should insist that the input be two-dimensional, but there is no need to specify types any more concretely than that.)

**Comment by timholy** [comment]
Closed by #577
